### DISCUSSION TITLE
Interface constructors

### DISCUSSION THREAD

**Discussion by UnoSD** [dotnet/csharplang#837] [General]
I was looking at this proposal from @devedse : https://github.com/dotnet/csharplang/issues/769

And it would be a nice language feature although, I agree with @jnm2 on his comment: https://github.com/dotnet/csharplang/issues/769#issuecomment-323382615 about the lack of semantical meaning

How about defining the constructor on the interfaces? That would solve both problems; it will allow parameters in constructors and the interface can specify the meaning with named parameters and a constructor that is bound to the interface meaning itself.

```csharp
class Program
{
    static void Main(string[] args)
    {
        var randomData1 = CreateRandomData<FastRandom>(10);
        var randomData2 = CreateRandomData<CryptoRandom>(10);
    }

    public static IEnumerable<int> CreateRandomData<R>(int length) where R : IRandom
    {
        var randomGenerator = new R(50);

        for (int i = 0; i < length; i++)
            yield return randomGenerator.Next();
    }
}

public interface IRandom
{
    IRandom(int seed);
    int Next();
}

public class FastRandom : IRandom
{
    private int seed;

    public FastRandom(int seed) => this.seed = seed;

    public int Next() => 0 + seed;
}

public class CryptoRandom : IRandom
{
    private int seed;

    public CryptoRandom(int seed) => this.seed = seed;

    public int Next() => 1 + seed;
}
```

Haven't thought this through yet so please apologise if it has some evident pitfall.

For instance, I, for one, struggle with the idea of forcing a parameter on an interface that has potentially no effect on the public API output but it is just an internal implementation detail.

**Comment by Pzixel** [comment]
IMHO it has nothing to do with interface. For example, how compiler have to choose between different constructors? Meanwhile, 
```csharp
public static IEnumerable<int> CreateRandomData<R>(int length) where R : IRandom, new(int)
```
 seems to be far more reasonable. And in this case IRandom is optional. For example:

```csharp
    public static IEnumerable<R> CreateRandomData<R>(int length) where R : new(int)
    {
        var randomGenerator = new Random(50);

        for (int i = 0; i < length; i++)
            yield return new R(randomGenerator.Next());
    }
```

**Comment by vladd** [comment]
In the typical case, classes implementing the same interface nevertheless need different parameters in their constructors.

```
public class NoRandom : IRandom
{
    public int Next() => 0;
}

public class FastRandom : IRandom
{
    int seed, multiplier;
    public FastRandom(int seed, int multiplier)
    {
        this.seed = seed;
        this.multiplier = multiplier;
    }
    public int Next() { seed *= multiplier; return seed; }
}

public class CombiningRandom : IRandom
{
    IRandom source, index;
    int[] reservoir = new int[256];
    public CombiningRandom(IRandom source, IRandom index)
    {
        this.source = source;
        this.index = index;
        for (int i = 0; i < reservoir.Length; i++) reservoir[i] = source.Next();
    }
    public int Next()
    {
        uint nextIdx = (uint)index.Next() >> 24;
        int result = reservoir[nextIdx];
        reservoir[nextIdx] = source.Next();
        return result;
    }
}
```

Having a predefined constructor signature would require squeezing the parameters into the signature types, which isn't going to produce a good code. Let's not endorse bad coding patterns.

**Reply by StellarWolfEntertainment** [reply]
Not all implementations of an interface will require different constructors, especially if you take into account potential generics, for example if you were to define this (I don't know why you would need to, it's just the first example I thought of)
```csharp
public interface IKeyValuePair<TKey, TValue>
{
    IKeyValuePair(TKey key, TValue value); // this constructor would apply to any combination of generics, and be *A* required constructor in any implementation of IKeyValuePair
}
```

**Reply by HaloFour** [reply]
This mixes contract with implementation, and only works if you're expecting exactly one implementation.  As soon as you have differing implementations with different dependencies this "required constructor" becomes a liability, in that you are now forced to support construction with incomplete arguments for a necessary implementation.

**Reply by StellarWolfEntertainment** [reply]
I suppose I tend to think in reverse, where it seems that some people 'define' an interface, then define the various implementations of that interface, I write the implementations and work backwards towards the interface merging common behaviors and data into baser types, so for me if I define 10 implementations that could all be indicated by a single interface and they all share a constructor signature in common, I'd like to add that constructor to the basest type I can, an example of a time where one might use a 'required' constructor is to ensure that there is a constructor for a type that is intended to load external data from someplace else, eg from a file, now we want to either ensure a 'file path', or a 'Stream' is provided to satisfy the intended purpose of the interface.

**Reply by CyrusNajmabadi** [reply]
>  like to add that constructor to the basest type I can

This is where an abstract class is appropriate. Note you're not mixing contract with implementation. The interface remains the contract, the abstract class can provide a common impl.

**Reply by HaloFour** [reply]
> I'd like to add that constructor to the basest type I can

Constructors aren't inherited.  Even if you did require a constructor in an abstract base class, and each derived class is required to call it, there is no requirement that each derived class expose a constructor with that same signature.  That's intentional, as if the language did require the same signature it'd be impossible (or extremely awkward) for any derived type to require more/different parameter than its base class.

**Comment by Opiumtm** [comment]
Most obvious and direct solution is to use "factory pattern".

Just define "factory interface"
```C#
public interface ISomething
{
    // ...
}

public interface ISomethingFactory<T> where T : ISomething
{
    T Create(int arg1, string arg2);
}
```

**Comment by jnm2** [comment]
What does this gain you over simply declaring a separate factory interface or custom delegate?
The separate factory interface makes more sense than including all the methods in one interface because the concerns are separate. You might want to use a caching factory in some places and a transient factory in other places while both factories return the same implementation.

Plus, like @vladd says there's the problem of different implementations needing extra constructor parameters which is almost a given.

**Reply by StellarWolfEntertainment** [reply]
There are plenty of times in which an implementation regardless of type might require otherwise identical constructors, I posted one example on the answer by @vladd, but there are other cases, an existing interface (but should admittedly be obsolete, since the system that used it is not advised for use, and itself has been obsoleted) would be the ISerializable interface where in order to properly implement it, it requires a constructor with a specific signature.

**Comment by Pzixel** [comment]
@jnm2 C# is a turing-complete language so you already can express anything, but it's better to have a language support than write factory of factory of adapter of facade. You wrote an entire factory just to invoke a constructor. It's a completly mechanical process and this is what compilers are good at.

**Comment by orthoxerox** [comment]
@jnm2 it's the same kind of request as the one about async constructors (#419). In both cases you can use a factory method, but bare factory methods are an antipattern in my opinion. You should construct objects using constructors, and it's an unnecessary limitation of the language that a constructor is a special method that always automatically allocates a new object. I would gladly welcome constructors that return cached instances, construct objects asynchronously etc.

Why can't the rules of object instantiation be a part of the interface? They are no different from the rules of interaction.

**Comment by HaloFour** [comment]
Constructors aren't virtual and aren't inherited behavior.  It generally doesn't make any sense for them to be given each type has very different dependencies.

That said, I would imagine that shapes would satisfy this proposal.

**Comment by Pzixel** [comment]
@HaloFour constructors has defenitly inherited behaviour. It has been discussed already: https://github.com/dotnet/csharplang/issues/419#issuecomment-294190226 and https://github.com/dotnet/roslyn/issues/6788#issuecomment-272162052 . I didn't invent any worthwhile solution but I defenitly think that we need more asyncish support in language and it's possible, but just not this obvious that it can be easily found.

Anyway, it's beside the point and proposed issue.

**Comment by HaloFour** [comment]
@Pzixel 

What you are describing is not inheritance.  Constructors are not inherited, which is why a derived class must invoke a parent constructor from its own constructor.

**Comment by Pzixel** [comment]
@HaloFour In this regard you're right, because I meant logical "inheritance".

**Comment by DavidArno** [comment]
I subscribe to the view that the need for a design pattern in code is an indication of a deficit in that language. Factory patterns are one such example. As @orthoxerox says,  `new` really is very limited in C#: it returns a new instance of some type. In situations where a value singleton is needed, a cached value, or some other control over what type is provided to the caller, then factories are needed. I do not subscribe to the view that factory methods are anti-patterns though. They are genuine patterns: they provide a work-around to the language's limitations.

However, I do not see this proposal as addressing any of this. What parameters a class needs for its constructors is implementation specific. Binding all implementations to a specific constructor defined by an interface feels like a step in the wrong direction to me.

**Comment by UnoSD** [comment]
Another downside is that it may encourage people to use it for logic classes forcing either to "service-locate" dependencies or to avoid DI all together (although it is not a good argument not to add features, see local functions...).

Perhaps we could solve this in a different fashion, as having to write a factory to achieve this is of course good practice, but a lot of boilerplate in simple cases. User could rely on high order functions passing the factory method in, but that would make it even closer to the implementation in some cases.

**Comment by jnm2** [comment]
I don't understand why you'd ever expect to be able to return arbitrary objects from a constructor. We need the concept of construction. It's unique; it *means* "build a brand new instance." You can't replace it with a higher-level factory abstraction. Any factory abstraction would have to delegate back to construction at some point so the factory abstraction itself wouldn't be called "construction."

**Comment by DavidArno** [comment]
@jnm2,
> I don't understand why you'd ever expect to be able to return arbitrary objects from a constructor.

I'm not sure we would expect to. `new` is just a built-in factory. But other factories are needed too, thus the various factory patterns. Language syntax for replacing those patterns would be nice. But it really isn't a priority in my view.

**Comment by iam3yal** [comment]
#164 should solve this problem.

**Comment by STeeL835** [comment]
Can someone point me to a relevant discussion/issue about this?

I have 20 classes of the same family, they are happy to have same ctor, but I don't want to create 20 different factories just to be able to instantiate them by passing arguments without change. Instead I use Activator, which is unsafer, because it's not static. Using a ServiceProvider to handle occasional parameter swaps is overkill imo.

**Reply by jnm2** [reply]
If you are able to use a type parameter that is constrained to your interface, then you could use a [static abstract interface method](https://learn.microsoft.com/en-us/dotnet/csharp/advanced-topics/interface-implementation/static-virtual-interface-members) named Create or similar and call `T.Create`.

**Reply by STeeL835** [reply]
Thanks, forgot about this. Still looks like ctor duplication, but at least not a factory in a separate file, I guess this will do for me.

**Comment by GeeLaw** [comment]
I think this is a solved issue?

```csharp
public interface IRandom
{
  abstract static IRandom Create(int seed);
  int Next();
}
public sealed class NotSoRandom : IRandom
{
  // Explicit to discourage creating instance with redundant argument
  // except in generics code.
  static IRandom IRandom.Create(int seed) => new NotSoRandom();
  public int Next() => 42;
}
```

Of course, the above isn't very safe if the implementing type is not `sealed`, in which case if a derived type forgets to reimplement the interface and redo the static creation methods, a wrong instance is created:

```csharp
public class Dangerous : IRandom
{
  private int seed_;
  public Dangerous(int seed) { seed_ = seed; }
  static IRandom IRandom.Create(int seed) => new Dangerous(seed);
  public int Next() => seed_;
}
public class Forgetful : Dangerous
{
  public Forgetful(int seed) : base(seed) { }
}
public static class Consumer
{
  public static TRandom CreateUltimateSeed<TRandom>()
    where TRandom : IRandom
  {
    // CreateUltimateSeed<Forgetful>() throws
    return (TRandom)TRandom.Create(42);
  }
}
```

The right way is hence to require knowing the actual type:

```csharp
public interface IRandom<TThis> : IRandom
  where TThis : IRandom<TThis>
{
  static IRandom IRandom.Create(int seed) => TThis.Create(seed);
  new abstract static TThis Create(int seed);
}
public class Dangerous : IRandom<Dangerous>
{
  private int seed_;
  public Dangerous(int seed) { seed_ = seed; }
  static Dangerous IRandom<Dangerous>.Create(int seed) => new Dangerous(seed);
  public int Next() => seed_;
}
public class Forgetful : Dangerous
{
  public Forgetful(int seed) : base(seed) { }
}
public class Remembering : Dangerous, IRandom<Remembering>
{
  public Remembering(int seed) : base(seed) { }
  // Forgetting the line below causes CS8705 (not most specific implementation)
  static IRandom IRandom.Create(int seed) => new Remembering(seed);
  static Remembering IRandom<Remembering>.Create(int seed) => new Remembering(seed);
}
public static class Consumer
{
  public static TRandom CreateUltimateSeed<TRandom>()
    where TRandom : IRandom<TRandom>
  {
    // CreateUltimateSeed<Forgetful>() fails to compile because it's not self-creatable
    return TRandom.Create(42);
  }
}
```

**Reply by STeeL835** [reply]
Oh, right, I forgot about static interface members. Still a bit redundant (basically 2 ctors), but at least no need to create another file, thanks!
